homeassistant/components/homekit/__init__.py

```diff
@@ -118,19 +118,19 @@ def get_accessory(hass, state, aid, config):
         unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
         device_class = state.attributes.get(ATTR_DEVICE_CLASS)
 
-        if device_class == DEVICE_CLASS_TEMPERATURE or unit == TEMP_CELSIUS \
-                or unit == TEMP_FAHRENHEIT:
+        if device_class == DEVICE_CLASS_TEMPERATURE and (
+                unit == TEMP_CELSIUS or unit == TEMP_FAHRENHEIT):
```

Comment:
The frontend doesn't check the `unit_of_measurement` anywhere so it would feel a bit weird to me if we did here.

Also, I think for temperature there's no ambiguity when using the unit of measurement. 

I think the better way to solve this whole problem is to do it like this:

 1. If a device class is set, use that for determining `a_type`
 2. If it's not set, try to understand the a_type from the unit of measurement.

Reply:
The unit must be either Celsius or Fahrenheit for the necessary conversion in the HomeKit type class. The reason I went with this quite strict approach is that the sensor type classes expect a specific sensor otherwise they might fail.

Reply:
Ok, I understand that, so `°C` or `°F` are required for unit of measurement.

However, I think at least with temperature, a quite common device type, we can do better than force users to set `device_class` to `temperature`. Not too many integrations use `device_class` yet and it will take some time until it's widely used.

Especially with unit of measurements like `°C` or `°F` it's quite clear that it's a `temperature` device class. With `%`, I agree that there's quite a bit of ambiguity and that's a problem that can be fixed by this PR. But I do think we can be more user-friendly with `°C` and `°F` units of measurement.

Reply:
The reason it did that for temperature as well was to do only one breaking change. However I see the point that it might make sense to wait a bit longer, until more sensors support it.

I will exclude temperature for now.